File: android/app/src/main/java/com/devloottask/gyroscopemodule/GyroscopeModule.java
package com.devloottask.gyroscopemodule;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

public class GyroscopeModule extends ReactContextBaseJavaModule implements SensorEventListener {

    private SensorManager sensorManager;
    private Sensor gyroscopeSensor;

    public GyroscopeModule(ReactApplicationContext reactContext) {
        super(reactContext);
        sensorManager = (SensorManager) reactContext.getSystemService(ReactApplicationContext.SENSOR_SERVICE);
        gyroscopeSensor = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE);
    }

    @Override
    public String getName() {
        return "GyroscopeModule";
    }

    @ReactMethod
    public void getGyroscopeData(final Callback callback) {
        if (gyroscopeSensor != null) {
            sensorManager.registerListener(this, gyroscopeSensor, SensorManager.SENSOR_DELAY_NORMAL);
        } else {
            callback.invoke("Gyroscope not available", null);
        }
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        sensorManager.unregisterListener(this); // Unregister listener after receiving data

        float[] rotationRate = event.values;

        double rotationRateX = rotationRate[0];
        double rotationRateY = rotationRate[1];
        double rotationRateZ = rotationRate[2];
        String orientation = calculateOrientation(rotationRateX, rotationRateY);

        WritableMap data = Arguments.createMap();
        data.putDouble("rotationRateX", rotationRateX);
        data.putDouble("rotationRateY", rotationRateY);
        data.putDouble("rotationRateZ", rotationRateZ);
        data.putString("orientation", orientation);

        sendEvent("GyroscopeData", data);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Not needed for gyroscope
    }

    private String calculateOrientation(double rotationRateX, double rotationRateY) {
        double angle = Math.atan2(rotationRateY, rotationRateX);
        double degrees = Math.toDegrees(angle);
        if (degrees >= -45 && degrees <= 45) {
            return "Landscape"; // Landscape
        } else {
            return "Portrait"; // Portrait
        }
    }

    private void sendEvent(String eventName, WritableMap params) {
        getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
}

File: ios/GyroscopeModule/GyroscopeModule.swift
//
//  GyroscopeModule.swift
//  DevlootTask
//
//  Created by Malik Shahbaz Haider on 28/02/2024.
//

import Foundation
import CoreMotion

@objc(GyroscopeModule)
class GyroscopeModule: NSObject {
    
    private let motionManager = CMMotionManager()
    
    @objc
    func getGyroscopeData(_ callback: @escaping RCTResponseSenderBlock) {
        guard motionManager.isGyroAvailable else {
            callback(["Gyroscope is not available on this device.", NSNull()])
            return
        }
        
        motionManager.startGyroUpdates(to: OperationQueue.main) { (gyroData, error) in
            guard let gyroData = gyroData, error == nil else {
                callback(["Failed to retrieve gyroscope data.", NSNull()])
                return
            }
            
            let rotationRateX = gyroData.rotationRate.x
            let rotationRateY = gyroData.rotationRate.y
            let rotationRateZ = gyroData.rotationRate.z
            let orientation = self.calculateOrientation(x: rotationRateX, y: rotationRateY)
                      
          
            let data = [
                "rotationRateX": rotationRateX,
                "rotationRateY": rotationRateY,
                "rotationRateZ": rotationRateZ,
                "orientation" : orientation
            ]
            
            callback([NSNull(), data])
        }
    }
  
  func calculateOrientation(x: Double, y: Double) -> String {
          let angle = atan2(y, x)
          let degrees = angle * (180.0 / Double.pi)
          if degrees >= -45 && degrees <= 45 {
              return "landscape"
          } else {
              return "portrait"
          }
      }
}
